docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -39,63 +43,208 @@ You can find more examples of taint rules in the [Semrep Registry](https://semgr
+      ...
+  - pattern: $X
+```
+
+This has two main limitations. First, it will override any sanitization that may be performed on the code matched by `$X`.  In the example code below, the call `sink(x)` will be reported as tainted despite `x` has been re-assigned to a fresh set!
+
+```python
+make_tainted(x)
+x = set([])
+sink(x) # false positive
 ```
 
-Mini cookbook
--------------
+Also, we must note that [`...` has limitations](/writing-rules/pattern-syntax/#ellipses-and-statement-blocks) and, for example, in the code below we would not get any finding from Semgrep if we used that source specification:
 
-Again, keep in mind that sources, sanitizers and sinks are given by arbitrary patterns, so they can be anything that you can match with Semgrep. You can get very creative!
+```python
+if cond:
+    make_tainted(x)
+sink(x) # false negative
+```
 
-### Function argument as a source
+The recommended solution nowadays is to set `by-side-effect: true`, being `by-side-effect` an option that is available for each individual source in a taint rule. When this option is enabled, _and_ the source specification matches a variable (or in general, an l-value) exactly, then Semgrep will assume that the variable (or l-value) becomes tainted by side-effect at the precise places where the source specification produces a match.
 
-Taint may come from specific functions that read user input such as `window.prompt()` but it is also easy to specify that taint comes from anywhere else, such as, for example, a specific argument within a function definition:
+So, this is how we specify that `make_tainted` will update its argument making future occurrences tainted:
 
-```javascript
+<iframe src="https://semgrep.dev/embed/editor?snippet=eRvy" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+There are also [sanitizers by side effect](#sanitizers-by-side-effect). For a good example of use of both sources and sanitizers by side-effect, take a look at rule [`c.lang.security.use-after-free.use-after-free`](https://semgrep.dev/playground/r/c.lang.security.use-after-free.use-after-free).
+
+Propagators
+-----------
+
+By default, tainted data automatically propagates through assignments, operators, and function calls (from inputs to output). However, there are other ways taint can propagate, which may require language- or library-specific knowledge that Semgrep does not have built in.
+
+Consider the following Python code where an unsafe `user_input` is stored into a `set` data structure. A random element from `set` is then passed into a `sink` function. This random element can be `user_input` itself, leading to an injection vulnerability!
+
+```python
+def test(s):
+    x = user_input
+    s = set([])
+    s.add(x)
+    #ruleid: test
+    sink(s.pop())
+```
+
+The following rule cannot find the above-described issue. The reason is that Semgrep does not know that executing `s.add(x)` makes `x` one of the elements in the set data structure `s`.
```

Comment:
```suggestion
The following rule cannot find the above-described issue. The reason is that Semgrep does not evaluate that executing `s.add(x)` makes `x` one of the elements in the set data structure `s`.
```

Reply:
Here `evaluate` sounds weird to me, I am proposing `is not aware`. Otherwise maybe something like "has no information about" is also OK.